docs: add Neon Auth integration guide (#49022)
expo/expo: docs/pages/guides/using-neon-auth.mdx

---
title: Using Neon Auth
description: Learn how to add Neon Auth for authentication in your Expo and React Native projects.
platforms: ['android', 'ios', 'web']
---

import { BoxLink } from '~/ui/components/BoxLink';
import { Prerequisites, Requirement } from '~/ui/components/Prerequisites';
import { Terminal } from '~/ui/components/Snippet';
import { Step } from '~/ui/components/Step';
import { Tabs, Tab } from '~/ui/components/Tabs';
import { BookOpen02Icon } from '@expo/styleguide-icons/outline/BookOpen02Icon';

> [Neon Auth](https://neon.com/docs/auth/overview) is a managed authentication service built on [Better Auth](https://www.better-auth.com/), the TypeScript-first, framework-agnostic authentication library. Neon Auth provides zero-config managed authentication that syncs users directly to your Neon Postgres database, giving you full ownership of your user data.

Reviewer: ✏️ **style** `prose.md#anti-patterns`

"Zero-config" and "full ownership of your user data" are the vendor's marketing, and the first one is contradicted by the four configuration steps that follow. State what the reader gets.

```suggestion
[Neon Auth](https://neon.com/docs/auth/overview) is a managed authentication service built on [Better Auth](https://www.better-auth.com/), the TypeScript-first, framework-agnostic authentication library. Neon Auth syncs users to your Neon Postgres database, so your user records live alongside your app data.
```

<!-- docs-review -->

> This guide shows you how to integrate Neon Auth with your Expo app using the `@neondatabase/auth` package.

Reviewer: ⚠️ **correctness** `comment-format.md#severity`

Say that this is beta. Neon's own overview states "The Managed Better Auth is in Beta", and every one of the 27 published versions of `@neondatabase/auth` is a prerelease, with `latest` currently at 0.5.0-beta. Name the version the guide targets so the reader knows the API can move under them.

<!-- docs-review -->

## Prerequisites

<Prerequisites>
  <Requirement title="Create a Neon account and project">

>     Sign up at [neon.com](https://neon.com) and create a project. Your authentication tables will be automatically created.

Reviewer: ✏️ **style** `prose.md#voice-and-tone`

Present tense and active voice: name who creates the tables.

```suggestion
    Sign up at [neon.com](https://neon.com) and create a project. Neon creates the authentication
    tables for you.
```

<!-- docs-review -->

  </Requirement>
  <Requirement title="Enable Neon Auth">
    In your Neon project dashboard, navigate to the Auth section and enable Neon Auth.
  </Requirement>
  <Requirement title="Use Expo SDK 53 or later">

>     Neon Auth works with Expo SDK 53 and later.

Reviewer: ⚠️ **correctness** `review-ai-text.md#t7-overclaiming`

Nothing supports this floor. `@neondatabase/auth` declares no `expo` or `react-native` peer dependency, ships no React Native entry point, and Neon's quick starts cover Next.js, React, and TanStack Router only. Give the reader the version you actually tested against, the way the Clerk requirement cites its peer dependency range.

<!-- docs-review -->

>   </Requirement>

Reviewer: ```suggestion
```

Not required. Currently, supported version is SDK 57 and deprecated versions 56, 55, and 54.

</Prerequisites>

## Install and configure Neon Auth

<Step label="1">

### Install the Neon Auth package

<Terminal
  cmd={{
    npm: ['$ npm install @neondatabase/auth'],
    yarn: ['$ yarn add @neondatabase/auth'],
    pnpm: ['$ pnpm add @neondatabase/auth'],
    bun: ['$ bun add @neondatabase/auth'],
  }}
/>

</Step>

<Step label="2">

### Add your Neon connection string

Add your Neon database URL to a **.env** file in your project root:

```bash .env
EXPO_PUBLIC_NEON_AUTH_URL=https://your-project.neon.com/auth

> DATABASE_URL=<your-neon-connection-string>

Reviewer: ⚠️ **correctness** `comment-format.md#severity`

Neon documents the auth URL as the database endpoint host, `https://ep-xxx.neon.tech/neondb/auth`, not `your-project.neon.com`. `DATABASE_URL` is never used anywhere on this page, and an Expo app has no server to read it, so asking the reader to paste a credentialed connection string into the app project is a risk with no payoff. Drop it here and drop the `DATABASE_URL` sentence from the callout below.

```suggestion
EXPO_PUBLIC_NEON_AUTH_URL=https://ep-example-123456.neon.tech/neondb/auth
```

<!-- docs-review -->

```

> > **info** The `EXPO_PUBLIC_` prefix is required because [Expo inlines these values at build time](/guides/environment-variables/#reading-environment-variables-from-env-files) so they are available in your JavaScript bundle. The auth URL is safe to expose as it's a public API endpoint. **Do not** add the `EXPO_PUBLIC_` prefix to `DATABASE_URL` as it contains credentials and is only used server-side.

Reviewer: ✏️ **style** `mdx-components.md#callouts`

The operative sentence here is a **Do not** instruction about credentials reaching the JavaScript bundle, so the keyword should be `warning`. The Clerk guide keeps the same explanation in plain prose if you would rather not use a callout at all.

<!-- docs-review -->

</Step>

<Step label="3">

### Create your auth configuration

Create a file to configure Neon Auth with your preferred authentication methods:

```tsx src/lib/auth.ts

> import { neonAuth } from "@neondatabase/auth";

Reviewer: ⚠️ **correctness** `comment-format.md#severity`

`@neondatabase/auth` does not export `neonAuth`. In 0.5.0-beta the root entry exports `createAuthClient`, `createInternalNeonAuth`, `AuthApiError`, `AuthError`, `isAuthApiError`, and `isAuthError`, and the package's own example is `createAuthClient('https://auth.example.com', { adapter: BetterAuthReactAdapter() })`. Social provider `clientId` and `clientSecret` are Better Auth server configuration, so they do not belong in an app file either.

<!-- docs-review -->

export const auth = neonAuth({
  emailAndPassword: {
    enabled: true,
  },
  socialProviders: {
    github: {
      clientId: process.env.GITHUB_CLIENT_ID!,
      clientSecret: process.env.GITHUB_CLIENT_SECRET!,
    },
    google: {
      clientId: process.env.GOOGLE_CLIENT_ID!,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET!,
    },
  },
});

export type Auth = typeof auth;
```

</Step>

<Step label="4">

### Wrap your app with the auth provider

In your root layout file, wrap your app with the Neon Auth provider:

```tsx src/app/_layout.tsx

> import { NeonAuthProvider } from "@neondatabase/auth/react";

Reviewer: ⚠️ **correctness** `comment-format.md#severity`

`@neondatabase/auth/react` does not export `NeonAuthProvider`. The provider it exports is `NeonAuthUIProvider`, and it takes different props than `authUrl`.

<!-- docs-review -->

import { Slot } from "expo-router";

const authUrl = process.env.EXPO_PUBLIC_NEON_AUTH_URL!;

export default function RootLayout() {
  return (
    <NeonAuthProvider authUrl={authUrl}>
      <Slot />
    </NeonAuthProvider>
  );
}
```

</Step>

## Add authentication UI

Use Neon Auth hooks to build your authentication screens:

<Tabs tabs={['Email/Password', 'Social Login']}>

<Tab>

```tsx src/app/sign-in.tsx

> import { useAuth } from "@neondatabase/auth/react";

Reviewer: ⚠️ **correctness** `comment-format.md#severity`

`@neondatabase/auth/react` does not export `useAuth`. The hooks it exports are `useAuthData`, `useAuthenticate`, `useCurrentOrganization`, `useStore`, and `useTheme`, and `signIn.email` is reached through the client returned by `createAuthClient`.

<!-- docs-review -->

        style={{ borderWidth: 1, padding: 10, marginBottom: 10 }}
      />
      <TextInput
        placeholder="Password"
        value={password}
        onChangeText={setPassword}
        secureTextEntry
        style={{ borderWidth: 1, padding: 10, marginBottom: 10 }}
      />
      {error && <Text style={{ color: "red" }}>{error}</Text>}
      <Button
        title={isLoading ? "Signing in..." : "Sign In"}
        onPress={handleSignIn}
        disabled={isLoading}
      />
    </View>
  );
}
```

</Tab>

<Tab>

```tsx src/app/sign-in.tsx
import { useAuth } from "@neondatabase/auth/react";
import { Button, View } from "react-native";

export default function SignInScreen() {
  const { signIn } = useAuth();

  return (
    <View style={{ padding: 20, gap: 10 }}>
      <Button
        title="Continue with GitHub"
        onPress={() => signIn.social({ provider: "github" })}
      />
      <Button
        title="Continue with Google"
        onPress={() => signIn.social({ provider: "google" })}
      />
    </View>
  );
}
```

</Tab>

</Tabs>

## Access user data

Use the `useSession` hook to access the authenticated user:

```tsx src/app/index.tsx

> import { useSession, useAuth } from "@neondatabase/auth/react";

Reviewer: ⚠️ **correctness** `comment-format.md#severity`

Neither `useSession` nor `useAuth` is exported from `@neondatabase/auth/react`. The package reads the session off the client instead: `const auth = createAuthClient(url, { adapter: BetterAuthReactAdapter() })`, then `auth.useSession()`.

<!-- docs-review -->

import { Button, Text, View } from "react-native";

export default function HomeScreen() {
  const { session, user, isLoading } = useSession();
  const { signOut } = useAuth();

  if (isLoading) {
    return <Text>Loading...</Text>;
  }

  if (!session) {
    return <Text>Please sign in</Text>;
  }

  return (
    <View style={{ padding: 20 }}>
      <Text>Welcome, {user?.name ?? user?.email}!</Text>
      <Button title="Sign Out" onPress={() => signOut()} />
    </View>
  );
}
```

## Query users in your database

Since Neon Auth syncs users directly to your Postgres database, you can query them alongside your app data:

```sql
-- Users are stored in the auth.users table
SELECT * FROM auth.users WHERE email = 'user@example.com';

-- Join with your app tables
SELECT 
  posts.*,
  auth.users.name as author_name
FROM posts

> JOIN auth.users ON posts.author_id = auth.users.id;

Reviewer: ⚠️ **correctness** `comment-format.md#severity`

Neon Auth stores its tables in the `neon_auth` schema, and the users table is `neon_auth.user`, singular. Every reference to `auth.users` here fails with an undefined-table error. Also say where the reader runs this, the way the Supabase guide titles its block "Supabase SQL Editor".

<!-- docs-review -->

```

## Run the app

> <Tabs tabs={['Expo Go', 'Android', 'iOS']}>

Reviewer: ✏️ **style** `terminology.md#expo-product-and-feature-names`

Expo Go is for testing a project, not for running it, and `npx expo start` is not specific to Expo Go, so the three labels are not variants of one thing. The Clerk guide has no such section, and starting an app is not what a reader comes to an authentication guide for, so consider dropping the section.

<!-- docs-review -->

<Tab>

<Terminal
  cmd={{
    npm: ['$ npx expo start'],
    yarn: ['$ yarn expo start'],
    pnpm: ['$ pnpm expo start'],
    bun: ['$ bun expo start'],
  }}
/>

</Tab>

<Tab>

<Terminal
  cmd={{
    npm: ['$ npx expo run:android'],
    yarn: ['$ yarn expo run:android'],
    pnpm: ['$ pnpm expo run:android'],
    bun: ['$ bun expo run:android'],
  }}
/>

</Tab>

<Tab>

<Terminal
  cmd={{
    npm: ['$ npx expo run:ios'],
    yarn: ['$ yarn expo run:ios'],
    pnpm: ['$ pnpm expo run:ios'],
    bun: ['$ bun expo run:ios'],
  }}
/>

</Tab>

</Tabs>

## Supported plugins

Neon Auth is a managed Better Auth service, so you don't install plugins yourself. A supported subset is exposed through the Neon SDK. See the [plugins guide](https://neon.com/docs/auth/guides/plugins) for configuration details.

> | Plugin | Status |

Reviewer: 🛑 **blocker** `mechanics.md`

This page is not formatted to the repository style, so it cannot land as is. Code samples use double quotes where the docs use single, `SELECT ` in the SQL block carries trailing whitespace, and this table's columns are unpadded. Run `pnpm format` in **docs**.

<!-- docs-review -->

| ------ | ------ |
| [Admin](https://neon.com/docs/auth/guides/plugins/admin) | Supported |
| [Email OTP](https://neon.com/docs/auth/guides/plugins/email-otp) | Supported |
| [JWT](https://neon.com/docs/auth/guides/plugins/jwt) | Supported |
| [Magic Link](https://neon.com/docs/auth/guides/plugins/magic-link) | Supported |
| [Organization](https://neon.com/docs/auth/guides/plugins/organization) | Partial |
| [Open API](https://neon.com/docs/auth/guides/plugins/openapi) | Supported |
| [Phone Number](https://neon.com/docs/auth/guides/plugins/phone-number) | Supported |

> > **info** MFA and additional plugins are on the [Neon Auth roadmap](https://neon.com/docs/auth/roadmap).

Reviewer: ✏️ **style** `prose.md#word-usage`

Spell out MFA on its only mention on this page.

```suggestion
> **info** Multi-factor authentication and additional plugins are on the [Neon Auth roadmap](https://neon.com/docs/auth/roadmap).
```

<!-- docs-review -->

## Next steps

<BoxLink
  title="Neon Auth documentation"

>   description="Complete guide to Neon Auth features including plugins, RBAC, and advanced configuration."

Reviewer: ✏️ **style** `prose.md#word-usage`

Spell out RBAC on its only mention on this page.

```suggestion
  description="Complete guide to Neon Auth features including plugins, role-based access control, and advanced configuration."
```

<!-- docs-review -->

  href="https://neon.com/docs/auth/overview"
  Icon={BookOpen02Icon}
/>

<BoxLink
  title="Supported plugins"
  description="See which Better Auth plugins Neon Auth supports, including Admin, Email OTP, JWT, Magic Link, Organization, and Phone Number."
  href="https://neon.com/docs/auth/guides/plugins"
  Icon={BookOpen02Icon}

> />

Reviewer: ✏️ **style** `review-ai-text.md#t5-restatement`

The "Supported plugins" section above already links this same page and names the same plugins in a table. Drop this card.

<!-- docs-review -->
